ambari-server/src/main/java/org/apache/ambari/server/agent/stomp/AgentHostDataHolder.java

```diff
@@ -77,21 +75,30 @@ public T initializeDataIfNeeded(Long hostId, boolean regenerateHash) throws Amba
    * event to listeners.
    */
   public void updateData(T update) throws AmbariException {
-    //TODO need optimization for perf cluster
-    updateLock.lock();
-    try {
-      initializeDataIfNeeded(update.getHostId(), true);
-      if (handleUpdate(update)) {
-        T hostData = getData(update.getHostId());
-        regenerateDataIdentifiers(hostData);
-        setIdentifiersToEventUpdate(update, hostData);
-        if (update.getType().equals(STOMPEvent.Type.AGENT_CONFIGS)) {
-          LOG.info("Configs update with hash {} will be sent to host {}", update.getHash(), hostData.getHostId());
-        }
-        STOMPUpdatePublisher.publish(update);
+    data.compute(update.getHostId(), (id, current) -> {
+      if (current == null) {
+        current = initializeData(id, true);
+      }
+      T updated;
+      try {
+        updated = handleUpdate(current, update);
+      } catch (AmbariException e) {
+        LOG.error("Error during handling update for host: {} and class {}", id, getClass().getName(), e);
+        throw new RuntimeException("Error during handling update for host: " + id + " and class: " + getClass().getName(), e);
```

Comment:
Again isn't this a behavior change, we are throwing a Runtime exception vs Ambari?